### PR TITLE
SKT-6: Implementation of Middleware in NestJS

### PR DESCRIPTION
**Pull request by prasharma0**
**Summary**
Implementation of Middleware in NestJS 

**Description**
- Created a global Middleware and a specific Middleware using the class method to verify the request that was sent were logged.
- Controllers where we can define our routes using the provided request that returns a specific value.

### CODE REVIEW

File: src/common/middleware/apitoken.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/* eslint-disable prettier/prettier */
+import { Injectable, NestMiddleware } from "@nestjs/common";
+import { NextFunction, Request, Response } from "express";
+
+@Injectable()
+export class ApiToken implements NestMiddleware {
+  use(req: Request, res: Response, next: NextFunction) {
+    console.log(req.headers);
```
Comment (bmnepali) [inline src/common/middleware/apitoken.ts]: Use the `Logger` class for logging instead.

File: src/app.module.ts [hunk 2]
```diff
@@ -13,7 +15,11 @@ const envFilePath: string = path.join(
...
 @Module({
   imports: [ConfigModule.forRoot({ envFilePath, isGlobal: true })],
-  controllers: [AppController],
+  controllers: [AppController, LoginController],
   providers: [AppService],
 })
-export class AppModule {}
+export class AppModule implements NestModule {
+  configure(consumer: MiddlewareConsumer) {
+    consumer.apply(ApiToken).forRoutes({ path: "*", method: RequestMethod.ALL });
```
Comment (bmnepali) [inline src/app.module.ts]: Looks like you are applying the middle for all paths, even for the login paths. I think you will have to apply the middleware for the paths in `AppController`. Please check.

**Review by bmnepali — CHANGES_REQUESTED** [review]

File: src/common/middleware/auth.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/* eslint-disable prettier/prettier */
+import { BadRequestException, Injectable, NestMiddleware } from "@nestjs/common";
+import { NextFunction, Request, Response } from "express";
+
+@Injectable()
+export class Logger implements NestMiddleware {
```
Comment (bhuwantm) [inline src/common/middleware/auth.ts]: Is this suppose to be authentication middleware or is it meant for loggin request.
If its for authentication, name it accordingly.

File: src/common/middleware/auth.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  use(req: Request, res: Response, next: NextFunction) {
+    if(req.headers["api_token"] === "my_token"){
```
Comment (bhuwantm) [inline src/common/middleware/auth.ts]: If the token is supposed to be auth token, then its better to use Authorization header.

File: src/common/middleware/home.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+/* eslint-disable prettier/prettier */
+import { BadRequestException, Injectable, NestMiddleware } from "@nestjs/common";
+import { NextFunction, Request, Response } from "express";
+
+@Injectable()
+export class homeLogger implements NestMiddleware {
```
Comment (bhuwantm) [inline src/common/middleware/home.ts]: Again if this is for logging the request to home page, remove the token checking mechanism.

File: src/common/middleware/login.module.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import { AppService } from "src/app.service";
+import { homeLogger } from "./home";
+
+@Module({
+  controllers: [AppController],
+  providers: [AppService],
+})
+export class LoginModule implements NestModule{
+    configure(consumer: MiddlewareConsumer) {
+        consumer.apply(homeLogger).forRoutes({ path: "/", method: RequestMethod.POST });
```
Comment (bhuwantm) [inline src/common/middleware/login.module.ts]: Is the homeLogger middleware for post request only?

**Comment by bhuwantm** [comment]
### Formatting Problems
There seems to be too many formatting problems, better format the code with prettier before commiting the files. Alternatively if you are using vscode you can use formatOnSave feature for automatic code formatting on each save.


### Module Separation
Separating out the codes as per module. Create a *.module.ts file for each module and load those modules. This will be better approach instead of loading controller and service for each module in main module.


### Code Structuring
Structure out for code properly, try to follow a fix pattern. For example in your PR you are using you are creating login.module.ts in middleware folder.


### Verbosity
Variable naming should be verbose so that its easier for other developer to understand. For example in your case you are using Logging for authenticating the token. If its for authentication you could have named the middleware as "authentication".


### Logical Separation
Separate out logic for Authentication, Logging, HomeLogging etc. In your PR you are checking for token in both Loggin and HomeLogging middleware. Instead of this each middleware should have specific purpose and you could chain it like:

`consumer.apply(Logger, Authentication, HomeLogger)`

For reference: [Multiple Middleware](https://docs.nestjs.com/middleware#multiple-middleware)

### Naming Convention
Its not necessary but its better to follow naming convention. For example in your PR you are naming class as homeLogger instead of HomeLogger

PS: Also we don't need too many module for starter kit, just a single module with appropriate structure and flow will be enough.

**Review by bhuwantm — CHANGES_REQUESTED** [review]

**Comment by bhuwantm** [comment]
Also we dont need too many module for starter kit, just a single module with appropriate structure and flow will enough.

File: src/common/users/users.controller.ts [hunk 7]
```diff
@@ -1,42 +1,13 @@
...
+
 @Controller("users")
 export class UsersController {
-  constructor(private blogService: BlogsService) {}
-  @Post()
+  @Get()
   userInfo(): string {
-    return "USER PAGE";
+    return "Could you please check the Request Logged from specific middleware?";
   }
```
Comment (bhuwantm) [inline src/common/users/users.controller.ts:8]: leave a new line here

File: src/common/users/middleware/specificMiddleware.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+import { Injectable, NestMiddleware } from "@nestjs/common";
+import { NextFunction, Request, Response } from "express";
+
+@Injectable()
+export class SpecificLogger implements NestMiddleware {
+  use(req: Request, res: Response, next: NextFunction) {
+    console.log("Request logged from specific Middleware...", req.url, req.headers);
+    next();
+  }
+}
```
Comment (bhuwantm) [inline src/common/users/middleware/specificMiddleware.ts]: rename the file to specificLogger.middleware.ts

**Review by bhuwantm — CHANGES_REQUESTED** [review]

**Review by bhuwantm — APPROVED** [review]

**Review by iampujan — APPROVED** [review]

**Pull request merged**